core/commands/filestore.go

```diff
@@ -0,0 +1,315 @@
+				errors = true
+				fmt.Fprintf(res.Stderr(), "%s\n", r.ErrorMsg)
+			} else {
+				fmt.Fprintf(res.Stdout(), "%s\n", r.FormatLong())
+			}
+		}
+		if errors {
+			res.SetError(fmt.Errorf("errors while displaying some entries"), cmds.ErrNormal)
+		}
+	},
+	Type: filestore.ListRes{},
+}
+
+var verifyFileStore = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Verify objects in filestore.",
+		LongDescription: `
+Verify objects in the filestore.
+
+If one or more <obj> is specified only verify those specific objects,
+otherwise verify all objects.
+
+The output is:
+
+<status> <hash> <size> <path> <offset>
+
+Where <status> is one of:
+ok:       the block can be reconstructed
+changed:  the contents of the backing file have changed
+no-file:  the backing file could not be found
+error:    there was some other problem reading the file
+missing:  <obj> could not be found in the filestore
+ERROR:    internal error, most likely due to a corrupt database
+
+For ERROR entries the error will also be printed to stderr.
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("obj", false, true, "Cid of objects to verify."),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		_, fs, err := getFilestore(req)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+		args := req.Arguments()
+		if len(args) > 0 {
+			out := perKeyActionToChan(args, func(c *cid.Cid) *filestore.ListRes {
+				return filestore.Verify(fs, c)
+			}, req.Context())
+			res.SetOutput(out)
+		} else {
+			next, err := filestore.VerifyAll(fs)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+			out := listResToChan(next, req.Context())
+			res.SetOutput(out)
+		}
+	},
+	PostRun: func(req cmds.Request, res cmds.Response) {
+		if res.Error() != nil {
+			return
+		}
+		outChan, ok := res.Output().(<-chan interface{})
+		if !ok {
+			res.SetError(u.ErrCast(), cmds.ErrNormal)
+			return
+		}
+		res.SetOutput(nil)
+		for r0 := range outChan {
+			r := r0.(*filestore.ListRes)
+			if r.Status == filestore.StatusOtherError {
+				fmt.Fprintf(res.Stderr(), "%s\n", r.ErrorMsg)
+			}
+			fmt.Fprintf(res.Stdout(), "%s %s\n", r.Status.Format(), r.FormatLong())
+		}
+	},
+	Type: filestore.ListRes{},
+}
+
+var dupsFileStore = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Print block both in filestore and non-filestore.",
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		_, fs, err := getFilestore(req)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+		ch, err := fs.FileManager().AllKeysChan(req.Context())
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+		reader, writer := io.Pipe()
+		go func() {
+			defer writer.Close()
+			for cid := range ch {
+				have, err := fs.MainBlockstore().Has(cid)
+				if err != nil {
+					res.SetError(err, cmds.ErrNormal)
+					return
+				}
+				if have {
+					fmt.Fprintf(writer, "%v\n", cid)
+				}
+			}
+		}()
+		res.SetOutput(reader)
```

Comment:
Don't use a reader for output like this. Send objects over a channel like we do in the other commands so that the api is easier to deal with.